Add clamav as secondary scanning mechanism

mfojtik opened the pull request:
This builds on top of https://github.com/openshift/image-inspector/pull/53 for providing clam results and @pweil- https://github.com/openshift/image-inspector/pull/45 for the clam support.

I tried to merge these two together, make the clam actuall scan files in the Docker image (or container if we point it to the container root fs). 

Some sample output:

```
# image-inspector -image docker.io/mfojtik/virus-test:latest -scan-type clamav -clam-socket /var/run/clamd.sock
2017/06/01 19:51:35 Pulling image docker.io/mfojtik/virus-test:latest
2017/06/01 19:51:39 Extracting image docker.io/mfojtik/virus-test:latest to /var/tmp/image-inspector-530387680
2017/06/01 19:51:43 ClamAv scan took 1s
2017/06/01 19:51:43 DEBUG: api.ScanResult{ImageName:"docker.io/mfojtik/virus-test:latest", ImageID:"", Results:[]api.Result{api.Result{Name:"clamav", ScannerVersion:"ClamAV 0.99.2/21723/Mon Jun 13 15:53:00 2016", Timestamp:time.Time{sec:63631936301, nsec:778170681, loc:(*time.Location)(0x8d9f40)}, Reference:"file:///virus", Description:"Eicar-Test-Signature", Summary:[]api.Summary(nil)}}}
```

(this actually found my hidden `virus` file I put inside the `mfojtik/virus-test:latest` image). Combining this with `-post-results-url` you can send the output of this test where you want to.

This also adds `/results` endpoint for the HTTP server where the compacted results are served as well (in case you don't want to use the POST-hook).

mfojtik commented:
@simon3z @enoodle @pweil- added `/results` endpoint for the webdav server so you will be able to get the clam results and compacted openscap results in old fashion as well.

mfojtik commented:
@pweil- i realized that we perhaps need something that indicates whether this is result of "container" scan or "image" scan. The difference between is that we don't have image ID for container scan.

mfojtik commented on pkg/clamav/scanner.go:
```diff
@@ -0,0 +1,86 @@
+	defer func() {
+		log.Printf("ClamAv scan took %ds", int64(time.Since(scanStarted).Seconds()))
+	}()
+	resultChan, err := scanner.ContScanFile(path)
```
@pweil- it seems like there are two options here, AllScanFile and ContScanFile, I was not able to tell the difference ;-)

pweil- replied:
ContScanFile is the directory one I believe.   That is what I expected to probably be using when we get past the WIP.

mfojtik replied:
@pweil- it works as expected, just wanted to make sure it is the right thing ;-) (i mean it gets me results and it finds the malicious file)

mfojtik commented on pkg/clamav/scanner.go:
```diff
@@ -0,0 +1,86 @@
+
+// ResultFileName returns the name of the results file
+func (s *ClamScanner) ResultsFileName() string {
+	return ""
```
@pweil- do we want/need to flush the clam results into disk? we serve them via HTTP endpoint (/results) and you can get them via post-hook.

mfojtik commented:
The last commit is trying to refactor the code internally to separate the scanners into their own options. Also plumb the image-inspector to not be openscap centric (there is still some work there to be done ;-)

mfojtik commented:
(the last commit can be a follow up, will have to plumb the tests...)

mfojtik commented on pkg/inspector/image-inspector.go:
```diff
@@ -35,6 +35,7 @@ const (
 	OWNER_PERM_RW            = 0600
 	HEALTHZ_URL_PATH         = "/healthz"
 	API_URL_PREFIX           = "/api"
+	RESULT_API_URL_PATH      = "/results"
```
@simon3z maybe this is not the best URL path but this will provide results for both openscap and clam... wonder if we want to version this api (i this it is a good idea to do).

mfojtik replied:
EDIT: I added `apiVersion` field to the JSON.

simon3z commented:
> @simon3z @enoodle @pweil- added /results endpoint for the webdav server so you will be able to get the clam results and compacted openscap results in old fashion as well.

@mfojtik nice that's perfect!

- Do we expect clamd to be running only on certain nodes and have these scans scheduled there?
- Are we planning to use a tcp socket (if possilbe) so that clamav could be a regular service in OpenShift?

mfojtik commented:
@simon3z i'm working on an example scanner pod with clam. The clamd can run inside the container and we can just point the image inspector to it's socket or we can have clamd on nodes and attach the socket via volume (maybe using the flex volume?)

service: we can run clam as http service (basically create proxy for the clamd socket).

simon3z commented:
> @simon3z i'm working on an example scanner pod with clam. The clamd can run inside the container and we can just point the image inspector to it's socket or we can have clamd on nodes and attach the socket via volume (maybe using the flex volume?)
> 
> service: we can run clam as http service (basically create proxy for the clamd socket).

Nice. The two containers + flex-volume seems good (also to keep the image-inspector container slim), but the http service (proxy to unix socket) seems really cool.

mfojtik commented:
OK, this is the example:

https://github.com/mfojtik/image-inspector/tree/clamav-example/examples/openshift/clamav

Basically:

```
[@dev] .../openshift/clamav # oc create -f clam-scan.yaml
pod "imagescan-1-rdrkz" created
[@dev] .../openshift/clamav # oc get pods
NAME                READY     STATUS              RESTARTS   AGE
imagescan-1-rdrkz   0/1       ContainerCreating   0          3s
[@dev] .../openshift/clamav # oc logs imagescan-1-rdrkz -f
LibClamAV Warning: **************************************************
LibClamAV Warning: ***  The virus database is older than 7 days!  ***
LibClamAV Warning: ***   Please update it as soon as possible.    ***
LibClamAV Warning: **************************************************
2017/06/05 12:02:23 clamd is ready to scan ...
2017/06/05 12:02:23 --> Waiting for scan results on 127.0.0.1:8080 ...
2017/06/05 12:02:23 Pulling image docker.io/mfojtik/virus-test:latest
2017/06/05 12:02:26 Extracting image docker.io/mfojtik/virus-test:latest to /image-data
2017/06/05 12:02:45 ClamAv scan took 15s
2017/06/05 12:02:45 DEBUG: Posting scan results to "http://127.0.0.1:8080/" ...
2017/06/05 12:02:45 --> Scan report received, 1 problems found ...
```

(the binary can do whatever, annotate the image, trigger something, upload something... endless possibilities ;-)

This demonstrate that you can run clam scan in Pod, however, the pod has to be privileged and ran as root (to access docker socket) and you have to mount in service account secret and you have to grant the service account to pull any image in the cluster (image-auditor role). We can fiddle the security of this if needed, but without image-inspector switching to containers/image for pulling we will have to run as root IMHO).

simon3z commented:
> This demonstrate that you can run clam scan in Pod, however, the pod has to be privileged and ran as root (to access docker socket) and you have to mount in service account secret and you have to grant the service account to pull any image in the cluster (image-auditor role).

@mfojtik really nice! Do you think you can keep the example in this repository? (As documentation)

mfojtik commented:
@simon3z sure

mfojtik commented:
@pweil- @simon3z @enoodle this is now ready :-)

simon3z commented:
@ilackarms @enoodle can you review?

enoodle commented on pkg/openscap/openscap.go:
```diff
@@ -210,7 +211,7 @@ func (s *defaultOSCAPScanner) Scan(mountPath string, image *docker.Image) error
 	_, err = s.chrootOscap(args...)
 
-	return err
+	return nil, err
```
Here we should read the result file and run `ParseResults` on it's contents?

enoodle commented on examples/openshift/clamav/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM openshift/origin:latest
```
Discussion: Should this clam deamon be part of image-inspector? If it is then we need more documentation on it either in the main README or in a local one.

mfojtik replied:
if we want to have one image supporting multiple scans, we need to think about adding more orchestration in the entrypoint (like have supervising process that run clam/etc)... agree, that is bigger discussion.

enoodle commented on README.md:
```diff
@@ -21,25 +21,46 @@ Image Inspector can extract docker images to a target directory and
+To build the image-inspector you can run this command:
 
-    $ godep go build
+    $ make
```
:+1:

enoodle commented on pkg/inspector/image-inspector.go:
```diff
@@ -203,12 +230,12 @@ func aggregateBytesAndReport(bytesChan chan int) {
 		case bytes, open := <-bytesChan:
 			if !open {
-				log.Printf("Finished Downloading Image (%dKb downloaded)", bytesDownloaded/1024)
+				log.Printf("DEBUG: Image download finished (%dKb downloaded)", bytesDownloaded/1024)
```
@simon3z Should these be debug messages? if they are, then we shouldn't print this always.

enoodle commented on pkg/cmd/types.go:
```diff
@@ -39,6 +39,10 @@ type ImageInspectorOptions struct {
+	// Local controls whether or not an image should be pulled from remote registry. If set
+	// to true, the image won't be pulled and instead the DstPath will be assumed as
+	// location to scan.
+	Local bool
```
Maybe this flag should have its own PR?

mfojtik replied:
yeah, I used this primarely for testing (you don't want to re-download the entire image every time ;-).

enoodle replied:
The docker client suppose to only download the image if it doesn't exist local or a new version is available

enoodle commented on pkg/inspector/image-inspector.go:
```diff
@@ -499,27 +526,28 @@ func (i *defaultImageInspector) getAuthConfigs() (*docker.AuthConfigurations, er
 }
 
-func (i *defaultImageInspector) scanImage(s iiapi.Scanner) ([]byte, []byte, error) {
+// TODO: Move this into openscap package
```
This function is meant to be used with all scanners, not just openscap.

mfojtik replied:
afaik this was just wrapping up openscap to get the report and html report... not every scanner will return these (like clam). guess we don't want to have specialized wrapper but rather honor the interface?

enoodle replied:
ok, then maybe we don't need this function here or to change the interface. This is just meant to reduce duplication from handling multiple scanners but as it was written with only one scanner it is obviously not sufficient.

mfojtik replied:
@enoodle i started to rewrite this then I realized that switching to interface will break tests ;-) but I think it is worth to do this right once and then we can re-use the interface.

mfojtik replied:
so I will do this refactor now, then use some `tig` magic to reduce scope for this PR so we can get this in smaller chunks hopefully.

simon3z commented on cmd/image-inspector.go:
```diff
@@ -18,24 +18,26 @@ func main() {
 	flag.StringVar(&inspectorOptions.DstPath, "path", inspectorOptions.DstPath, "Destination path for the image files")
 	flag.StringVar(&inspectorOptions.Serve, "serve", inspectorOptions.Serve, "Host and port where to serve the image with webdav")
 	flag.BoolVar(&inspectorOptions.Chroot, "chroot", inspectorOptions.Chroot, "Change root when serving the image with webdav")
+	flag.BoolVar(&inspectorOptions.Local, "local", inspectorOptions.Local, "If set the image wont be pulled but instead the -path will be used")
```
@mfojtik if this is related to scanning (e.g. scan local path) shouldn't contain a scan prefix? e.g. `scan-local-path` or something similar.

mfojtik replied:
this option means don't pull images trough docker and don't extract it (so do a local scan)... `scan-local` ?

simon3z replied:
:+1:

mfojtik replied:
done.

simon3z commented on examples/openshift/clamav/clam-scan.yaml:
```diff
@@ -0,0 +1,30 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: imagescan-1-rdrkz
```
@mfojtik should this use `generateName`?

mfojtik replied:
yeah, sorry was quick copy&paste ;-)

mfojtik commented:
@enoodle added clam tests (coverage: 95.8% of statements)

mfojtik commented:
@enoodle @pweil- @simon3z changes:

* addressed review comments
* refactored the Scanner interface a little (Scan() can now return interface{} which can hold additional results (arf + html for openscap) so there is no need for scanImage() wrapper)
* added clam unit test
* remove --local/scan-local flag for now (it will be separate PR)
* removed DEBUG: in download messages (can be discussed separately)

One remaining question here is the `/results` endpoint that serves the compacted JSON results. Do you think that URL path is ok?

mfojtik commented on pkg/api/types.go:
```diff
@@ -96,12 +97,11 @@ type APIVersions struct {
-	// ScannerName is the scanner's name
-	ScannerName() string
-	// ResultFileName returns the name of the results file
-	ResultsFileName() string
```
@enoodle @pweil- also no need for these... the Scanner interface should just Scan :-)

mfojtik commented:
@enoodle @pweil- @simon3z the last commit will cause that we will run the clamd as part of the scan if the `-clam-socket` is not specified. That allows to run this in container without doing complex process management. I would still like to have the `-clam-socket` option for complex scanning scenarios (custom config, shared clamd, clamd running directly on node, etc..)

simon3z commented:
> @enoodle @pweil- @simon3z the last commit will cause that we will run the clamd as part of the scan if the -clam-socket is not specified.

@mfojtik any chance we can keep this decoupled in another PR? I am not sure about it... I am trying to keep the image as slim as possible (in fact it would be great to have OpenSCAP outside as well).

I have several comments on this new part but I'd keep them for the specific PR.

mfojtik commented:
@simon3z np, I can keep it separate.

mfojtik commented:
@simon3z @enoodle @pweil- are there any other comments for this PR or ok to merge?

pweil- commented:
none from me except....squash the clam related ones?

mfojtik commented:
squashed, will wait for Federico

simon3z commented:
> squashed, will wait for Federico

@mfojtik I'll try to review to details this in the next 1-2 days is that OK?

mfojtik commented:
@simon3z perfectly ok, thanks!

mfojtik commented:
@joelsmith FYI

simon3z commented:
OK LGTM :+1:
Waiting for @joelsmith review.

joelsmith commented on Godeps/Godeps.json:
```diff
@@ -10,6 +10,9 @@
 			"ImportPath": "github.com/antchfx/xpath",
 			"Rev": "1f3266e7730781347c9decd229c1fe1c18be760c"
 		},
+			"ImportPath": "github.com/dutchcoders/go-clamd",
```
This project doesn't currently support clamd's FILDES method which passes a file descriptor across the socket using the `SCM_RIGHTS` message type. Because of this and because our clamd will not be able to scan by paths (since it's in a different fs namespace), it's going to have to stream the data across the socket (necessitating another copy of the data). This is something we were hoping to avoid. I think we should either:

1. Send a PR to the project that adds FILDES support
or
2. Use https://github.com/Miciah/clam-scanner instead. It was purpose built by us to solve this problem (since when I looked for a golang clamd front-end, I failed to find anything). You can see the FILDES code in action here:

https://github.com/Miciah/clam-scanner/blob/f34ad6de7f7afff39452e0e11f8549939e067b06/pkg/clamav/session.go#L317-L326

His `ScanPath()` function uses filepath.Walk() to scan an entire sub-tree.

mfojtik replied:
@joelsmith why you have to work in different fs namespace? is that for scanning live containers from the clam running on the host?

I would certainly prefer 1) and just add support for FILDES to go-clamd. Also I think we don't want to do FILDES scan by default as I don't  think scanning files in different FS namespace is a common use-case.

How long down the filepath.Walk() scan takes on large images?

pweil- replied:
On one hand the existing project already supports both tcp and socket connections as well as a variety of existing scan methods that we'd may want to support anyway (or maybe not, see hand number 3).  On the other hand there isn't enough activity that would make me think it already has strong adoption.  On the third hand, as long as we define the interface in image-inspector on how we expect to interact with the scanner I don't care which dependency we compile in and what it does in the underlying implementation when we give it a path (walk and stream or filedes).

So, do we want to maintain a new version, enhance the existing one by adding a `ScanFileDes` to the [existing scan code](https://github.com/dutchcoders/go-clamd/blob/master/clamd.go), or fork?

Both projects are small enough.

joelsmith commented on examples/openshift/clamav/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+ENTRYPOINT ["/bin/sh", "-c"]
+
+RUN yum install -y clamav-server clamav-scanner clamav-data --disablerepo=origin-local-release \
+      && mkdir -p /image-data && chmod 0777 /image-data
```
While it's great that you provide an example clam container, it's not really useful as anything more than a starting point since it would only have the very old signatures provided in the `clamav-data` package. It may be more useful to point at a more full-featured implementation in addition to this. We're currently building a clamd pod that will be run as a daemonset on our nodes. One container in the pod runs clamd and another periodically gets updates via freshclam and other sources as well as providing a hook to have cluster-specific signatures passed in. More of an FYI than anything else. I'll send you a link to it once we're a bit further along. In the mean time, it might be worth noting the deficiencies so nobody goes off and builds something around it only to discover that they're using old signatures with no good way of updating them.

mfojtik replied:
this PR was not meant to solve all problems, just give people starting point.

you can update clam signatures everytime when the image-inspector image starts (as part of the entrypoint). 

i think the disconnect between this and what this is trying to prove is that you are building something that will run as service on every node (via daemon-set) where this is trying to do the "job-based" approach, where you fire up a scan on demand (basically create a "job" with target image to scan and URL where it should post the result).

I dunno our infra details, but I would question why the clamd has to run on every node vs. having specialized nodes intended to do only scanning.

pweil- replied:
specialized nodes won't account for existing, running containers.  But that also doesn't preclude the use of a job based approach.  I don't necessarily think it's an either/or question in the long run but we probably want to stick with one approach to begin with.

simon3z replied:
Let's also remember that there is another degree of freedom: image-inspector as a library. If we design the code correctly then we wouldn't prevent anyone from building a long-running node agent `image-inspector-daemon` (of course it will need an extra amount of orchestration).
Maybe the daemon could be specialized only to scan existing containers so that it won't have to deal with storage management.
Anyway overall I still like the job-based approach, but I think we should facilitate anyone who would like to take a different approach.

mfojtik commented:
@joelsmith @pweil- @simon3z updated this to use the openshift clam-scanner package... couple follow ups:

1) openshift clam-scanner need to learn how to report versions
2) we should remove glog from the clam-scanner dependency or we asking for trouble (using two loggers)

PTAL

mfojtik commented:
@joelsmith @pweil- @simon3z are we ok with this version? (not using go-clamd)

follow-ups: 

* report the clam version properly (should be handled upstream)
* remove glog from clam scanner (to reduce num of dependencies)

pweil- commented:
no objections from me.  @joelsmith how do you feel about this direction?

joelsmith commented:
Thanks for doing all of this work. 👍

mfojtik commented:
@simon3z merge? or waiting in your queue? :)

simon3z commented:
> @simon3z merge? or waiting in your queue? :)

@mfojtik I'd feel more comfortable merging if I had explicit approvals from @enoodle and @pweil-

enoodle commented:
The Clam part LGTM. I am afraid that something got broken with the openscap without me noticing. I wanted to build and run it but for some reason `docker build` fails on the yum commands. I will update when I figure it out.

ilackarms commented:
update dockerfile? @mfojtik

mfojtik commented:
@ilackarms which one? the image-inspector one? i don't think we want to install clamd there.
